pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -568,7 +568,7 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 			DevicePath: devicePath,
 		}
 
-		if volumeDeviceMounter != nil {
+		if volumeDeviceMounter != nil && actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) != DeviceGloballyMounted {
```

Comment:
Is it possible to add unit tests for these changes?

Reply:
operation_generator.go does not have a big unit test coverage, I'll try to test it manually with a mock driver + some error injection.

Reply:
@jsafrane we have azure file csi driver that has implemented `NodeStageVolume`, and for this code change, it only call one `NodeStageVolume`, it would not work if that volume is going to mount on multiple nodes, shall we revert this code change? find details here: https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/620

Reply:
I am not sure I understand. The code above makes sure that `NodeStageVolume` should be called only once on each node where a shared volume is used. The volume can still be used in multiple nodes.

Reply:
from what I observe, there is a little possibility that `NodeStageVolume` is not invoked, only invoke `NodePublishVolume`, and it only happens from k8s v1.20.x, do you know what could be the root cause? thanks.
```
I0502 14:51:38.501550       1 utils.go:118] GRPC call: /csi.v1.Node/NodePublishVolume
I0502 14:51:38.501574       1 utils.go:119] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount","target_path":"/var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":1}},"volume_context":{"csi.storage.k8s.io/ephemeral":"false","csi.storage.k8s.io/pod.name":"azurefile-volume-tester-2tpth","csi.storage.k8s.io/pod.namespace":"azurefile-4885","csi.storage.k8s.io/pod.uid":"ed800670-69de-4a1e-902c-e0ceeca4ec68","csi.storage.k8s.io/serviceAccount.name":"default"},"volume_id":"kubetest-u33fnt21#f7bf7d45cebcf4286987d85#pre-provisioned-multiple-pods#"}
I0502 14:51:38.501732       1 nodeserver.go:84] NodePublishVolume: mounting /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount at /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount with mountOptions: [bind]
I0502 14:51:38.501750       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount)
E0502 14:51:38.503747       1 mount_linux.go:179] Mount failed: exit status 32
Mounting command: mount
Mounting arguments:  -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount
Output: mount: /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount: special device /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount does not exist.

E0502 14:51:38.503833       1 utils.go:123] GRPC error: rpc error: code = Internal desc = Could not mount "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount" at "/var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount": mount failed: exit status 32
Mounting command: mount
Mounting arguments:  -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount
Output: mount: /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount: special device /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount does not exist.
```

Reply:
I think I found the issue, there are multiple pods referencing same volume in the test case, those pods would be created, deleted in serial, so usually it won't have any issue, every pod creation and termination would have corresponding `NodeStageVolume` and then `NodePublishVolume` one by one; the error case could be `NodeUnstageVolume` takes some time(delay), and for next pod, it could be still in `DeviceGloballyMounted`, and then skip `NodeStageVolume` and after a while, the former `NodeUnstageVolume` completed, so for next pod, it only has `NodePublishVolume` func, is that possible @jsafrane

Reply:
another common error case: if two csi driver PVs has the same `volumeHandle` value, `NodeStageVolume` will only happen once and the same PV mount won't have `NodeStageVolume`, it would fail.

Reply:
If they have the same `volumeHandle`, they're the same volumes and hence a single NodeStage is called. From CSI `volume_id` description:
```
  // This field MUST contain enough information to uniquely identify
  // this specific volume vs all other volumes supported by this plugin.
  // This field SHALL be used by the CO in subsequent calls to refer to
  // this volume.
  // The SP is NOT responsible for global uniqueness of volume_id across
  // multiple SPs.
  string volume_id = 2;
```

Reply:
> `NodeUnstageVolume` takes some time(delay), and for next pod, it could be still in `DeviceGloballyMounted`, and then skip `NodeStageVolume` and after a while, the former `NodeUnstageVolume` completed, so for next pod, it only has `NodePublishVolume` func, is that possible @jsafrane

It should not be possible. Kubelet should run only a single operator per volume (MountVolume or UnmountVolume). The aforementioned behavior would be a bug, if it really happens.

Reply:
This affects the following scenario, can you please share your thoughts:

If there is a pod(let's say part of a StatefulSet) running on a node which uses a volume and the pod is deleted, the `NodeUnstageVolume` RPC gets called.

If completion of this `NodeUnstageVolume` RPC call takes time(for example: due to a wait on completion on `umount` Linux command), for the new replacement pod(scheduled on the same node) kubelet will find the volume as `DeviceGloballyMounted` and calls `NodePublishVolume` RPC without calling `NodeStageVolume` RPC. 

Later the `NodeUnstageVolume` RPC's unmount operation which was in progress would complete and unmount the volume. Thus, effectively for the new pod the staging doesn't happen.

Reply:
@rohitChaware did you hit such race condition issue recently?

Reply:
Yes @andyzhangx , reproduced the above scenario on K8s version 1.20.6. 

`NodeStageVolume` RPC isn't getting called in case when `NodeUnstageVolume` operation takes longer to complete(more than 2 minutes) and a new pod using the same volume is scheduled on the same node.

The flow of RPC calls go as expected for versions < 1.20.

Reply:
https://github.com/kubernetes/kubernetes/pull/100183 may help